catboost/libs/fstr/shap_values.cpp

```diff
@@ -1235,6 +1441,601 @@ TVector<TVector<double>> CalcShapValues(
+        classIndicesFirst.end(),
+        classIndicesSecond.begin(),
+        classIndicesSecond.end(),
+        std::back_inserter(sameClassIndices)
+    );
+    return sameClassIndices;
+}
+
+static inline void FillIndices(const TVector<size_t>& indices, TVector<bool>* isIndices) {
+    for (size_t idx : indices) {
+        isIndices->at(idx) = true;
+    }
+}
+
+static void ContructClassIndices(
+    const TVector<TVector<int>>& combinationClassFeatures,
+    const TMaybe<std::pair<int, int>>& pairOfFeatures,
+    TVector<size_t>* classIndicesFirst,
+    TVector<size_t>* classIndicesSecond
+) {
+    const size_t classCount = combinationClassFeatures.size();
+    if (pairOfFeatures.Defined()) {
+        for (size_t classIdx : xrange(classCount)) {
+            for (int flatFeatureIdx : combinationClassFeatures[classIdx]) {
+                if (flatFeatureIdx == pairOfFeatures->first) {
+                    classIndicesFirst->emplace_back(classIdx);
+                }
+                if (flatFeatureIdx == pairOfFeatures->second) {
+                    classIndicesSecond->emplace_back(classIdx);
+                }
+            }
+        }
+        if (classIndicesFirst->size() > classIndicesSecond->size()) {
+            std::swap(classIndicesFirst, classIndicesSecond);
+        }
+    } else {
+        classIndicesFirst->resize(classCount);
+        classIndicesSecond->resize(classCount);
+        Iota(classIndicesFirst->begin(), classIndicesFirst->end(), 0);
+        Iota(classIndicesSecond->begin(), classIndicesSecond->end(), 0);
+    }
+}
+
+template <typename TStorageType>
+static inline void AddInteractionEffectAllDocumentsForPairClasses(
+    const TVector<TVector<TVector<double>>>& contribsOn,
+    const TVector<TVector<TVector<double>>>& contribsOff,
+    size_t classIdx1,
+    size_t classIdx2,
+    int documentCount,
+    int approxDimension,
+    bool isSecondFeature,
+    bool isCalcMainEffect,
+    TStorageType* shapInteractionValuesInternal
+) {
+    if (classIdx2 == classIdx1) {
+        return;
+    }
+    auto valuesBetweenDifferentClasses = isSecondFeature ? GetDocumentsByClasses(shapInteractionValuesInternal, classIdx1, classIdx2) : nullptr;
+    auto valuesBetweenSameClasses = isCalcMainEffect ? GetDocumentsByClasses(shapInteractionValuesInternal, classIdx1, classIdx1) : nullptr;
+
+    for (int dimension : xrange(approxDimension)) {
+        for (size_t documentIdx : xrange(documentCount)) {
+            double interactionEffect = GetInteractionEffect(
+                contribsOn,
+                contribsOff,
+                documentIdx,
+                dimension,
+                classIdx2
+            );
+            if (isSecondFeature) {
+                (*valuesBetweenDifferentClasses)[dimension][documentIdx] = interactionEffect;
+            }
+            if (isCalcMainEffect) {
+                (*valuesBetweenSameClasses)[dimension][documentIdx] -= interactionEffect;
+            }
```

Comment:
after you change indexing in contribsOn and contribsOff, create array refs for contribsOn[classIdx2][dimension], contribsOff[classIdx2][dimension], (*valuesBetweenDifferentClasses)[dimension], (*valuesBetweenSameClasses)[dimension] before the inner loop and use them inside the loop.

you will also need to remove dimension and classIdx2 from the list of parameters of GetInteractionEffect